webrender/src/device/gfx/device.rs

```diff
@@ -1492,43 +1648,155 @@ impl<B: hal::Backend> Device<B> {
             .expect("Program not found.");
         let mut cmd_buffer = self.command_pool[self.next_id].acquire_command_buffer();
         unsafe { cmd_buffer.begin() };
-        let shader_group = program.shader_kind.into();
-        let need_alloc = {
-            let bound_resources = DescriptorSetResources::new(
-                shader_group,
-                [
-                    self.bound_textures[SAMPLERS[0].0],
-                    self.bound_textures[SAMPLERS[1].0],
-                    self.bound_textures[SAMPLERS[2].0],
-                    self.bound_textures[SAMPLERS[3].0],
-                    self.bound_textures[SAMPLERS[4].0]
-                ],
-            );
+        let descriptor_group = program.shader_kind.into();
+        // Per draw textures
+        let bound_resources = PerDrawResources::new(
+            [
+                self.bound_textures[SAMPLERS[0].0],
+                self.bound_textures[SAMPLERS[1].0],
+                self.bound_textures[SAMPLERS[2].0],
+            ],
+        );
 
-            let (desc_set, need_alloc) = match self.per_draw_descriptor_bindings[self.next_id].entry(bound_resources) {
-                Entry::Occupied(o) => {
-                    (self.descriptor_pools_per_draw[self.next_id].get_set_at_location(shader_group, *o.get()), false)
+        match self.per_draw_descriptor_bindings.entry(bound_resources) {
+            Entry::Occupied(_) => {
+                for &(index, _) in SAMPLERS[0..PER_DRAW_SAMPLER_COUNT].iter() {
+                    program.bound_textures[index] = self.bound_textures[index];
+                    let image = &self.images[&self.bound_textures[index]].core;
+                    // We need to transit the image, even though it's bound in the descriptor set
+                    let mut src_stage = Some(hal::pso::PipelineStage::empty());
+                    if let Some(barrier) = image.transit(
+                        hal::image::Access::SHADER_READ,
+                        hal::image::Layout::ShaderReadOnlyOptimal,
```

Comment:
I wonder if there needs to be any transition here at all. I'd assume that for most resources that are part of descriptor sets being `ShaderReadOnlyOptimal` is a normal state to be in. If, say, an image needs to receive some buffer transfers, it would change the layout, accept the transfer, and then change back the layout to `ShaderReadOnlyOptimal`. So by the time we reach here, we just need to assert that an image is in the proper state.